### ISSUE TITLE
PySAM makes 5 API calls to get the csv weather file

### ISSUE THREAD

**Issue by ghazi-git** [NREL/pysam#95]
I'm not sure if this is the expected behavior or I'm doing sth wrong but while debugging #94, I found out that pysam makes 5 API calls to get the .csv weather file. This has been happening consistently (it's weird that even the 429 response is received twice consistently before returning a 302). It would be great to avoid the extra calls and do just one (if possible), as that will improve the time needed to get the file.

Here are the API calls involved at a high-level:

- API Call 1:
    - **Request**: GET https://developer.nrel.gov/api/nsrdb/v2/solar/psm3-tmy-download.csv
    - **Response**: 429 Too Many Requests
- API Call 2 (Retrying the first API call):
    - **Request**: GET https://developer.nrel.gov/api/nsrdb/v2/solar/psm3-tmy-download.csv
    - **Response**: 429 Too Many Requests
- API Call 3 (Retrying the first API call):
    - **Request**: GET https://developer.nrel.gov/api/nsrdb/v2/solar/psm3-tmy-download.csv
    - **Response**: 302 Found (location: https://mapfiles.nrel.gov/data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv)
- API Call 4:
    - **Request**: GET https://mapfiles.nrel.gov/data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv
    - **Response**: 301 Moved Permanently (Location: https://gds-hsds-nsrdb-files.nrelcloud.org/data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv)
- API Call 5:
    - **Request**: GET https://gds-hsds-nsrdb-files.nrelcloud.org/data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv
    - **Response**: 200 OK

<details>
<summary>Here's the debug output for all the process of getting the weather files: json and csv (please note that I removed the warnings linked to #89 from the output) </summary>

```
Starting data download for solar using 1 thread workers.
Getting list of available NSRDB files for 43.0402873, -76.2002669.


DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): developer.nrel.gov:443
send: b'GET /api/nsrdb/v2/solar/nsrdb-data-query.json?api_key=API_KEY&wkt=POINT(-76.2002669%2043.0402873) HTTP/1.1\r\nHost: developer.nrel.gov\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n'
reply: 'HTTP/1.1 200 OK\r\n'
header: Date: Tue, 27 Jul 2021 09:49:46 GMT
header: Content-Type: application/json; charset=utf-8
header: Transfer-Encoding: chunked
header: Connection: keep-alive
header: Vary: Accept-Encoding
header: X-RateLimit-Limit: 1000
header: X-RateLimit-Remaining: 996
header: x-dns-prefetch-control: off
header: x-frame-options: SAMEORIGIN
header: strict-transport-security: max-age=63072000; includeSubDomains; preload
header: x-download-options: noopen
header: x-content-type-options: nosniff
header: x-xss-protection: 1; mode=block
header: content-security-policy: default-src 'self'
header: x-content-security-policy: default-src 'self'
header: x-webkit-csp: default-src 'self'
header: referrer-policy: same-origin
header: etag: W/"4341-frpGrOJc8Nn2Wkh2N+PyekpUuDE"
header: Age: 0
header: Via: https/1.1 api-umbrella (ApacheTrafficServer [cMsSf ])
header: X-Cache: MISS
header: Access-Control-Allow-Origin: *
header: Access-Control-Allow-Methods: POST, PUT, DELETE, GET, OPTIONS
header: Access-Control-Request-Method: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
header: Strict-Transport-Security: max-age=31536000; preload
header: Content-Encoding: gzip
DEBUG:urllib3.connectionpool:https://developer.nrel.gov:443 "GET /api/nsrdb/v2/solar/nsrdb-data-query.json?api_key=API_KEY&wkt=POINT(-76.2002669%2043.0402873) HTTP/1.1" 200 None
List of available data saved to /tmp/pysam/nsrdb_data_query_response_43.0402873_-76.2002669.json.


https://developer.nrel.gov/api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false
send: b'GET /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false HTTP/1.1\r\nHost: developer.nrel.gov\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n'
reply: 'HTTP/1.1 429 Too Many Requests\r\n'
header: Date: Tue, 27 Jul 2021 09:49:47 GMT
header: Content-Type: text/csv
header: Transfer-Encoding: chunked
header: Connection: keep-alive
header: X-RateLimit-Limit: 5000
header: X-RateLimit-Remaining: 4996
header: Access-Control-Allow-Origin: *
header: X-Cache: MISS
header: Access-Control-Allow-Methods: POST, PUT, DELETE, GET, OPTIONS
header: Access-Control-Request-Method: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
header: Strict-Transport-Security: max-age=31536000; preload
DEBUG:urllib3.connectionpool:https://developer.nrel.gov:443 "GET /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false HTTP/1.1" 429 None


DEBUG:urllib3.util.retry:Incremented Retry for (url='/api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false'): Retry(total=9, connect=10, read=10, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Retry: /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false
send: b'GET /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false HTTP/1.1\r\nHost: developer.nrel.gov\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n'
reply: 'HTTP/1.1 429 Too Many Requests\r\n'
header: Date: Tue, 27 Jul 2021 09:49:47 GMT
header: Content-Type: text/csv
header: Transfer-Encoding: chunked
header: Connection: keep-alive
header: X-RateLimit-Limit: 5000
header: X-RateLimit-Remaining: 4996
header: Access-Control-Allow-Origin: *
header: X-Cache: MISS
header: Access-Control-Allow-Methods: POST, PUT, DELETE, GET, OPTIONS
header: Access-Control-Request-Method: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
header: Strict-Transport-Security: max-age=31536000; preload
DEBUG:urllib3.connectionpool:https://developer.nrel.gov:443 "GET /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false HTTP/1.1" 429 None


DEBUG:urllib3.util.retry:Incremented Retry for (url='/api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false'): Retry(total=8, connect=10, read=10, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Retry: /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false
send: b'GET /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false HTTP/1.1\r\nHost: developer.nrel.gov\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n'
reply: 'HTTP/1.1 302 Found\r\n'
header: Date: Tue, 27 Jul 2021 09:49:49 GMT
header: Content-Type: text/plain; charset=utf-8
header: Content-Length: 125
header: Connection: keep-alive
header: X-RateLimit-Limit: 5000
header: X-RateLimit-Remaining: 4996
header: x-dns-prefetch-control: off
header: x-frame-options: SAMEORIGIN
header: strict-transport-security: max-age=63072000; includeSubDomains; preload
header: x-download-options: noopen
header: x-content-type-options: nosniff
header: x-xss-protection: 1; mode=block
header: content-security-policy: default-src 'self'
header: x-content-security-policy: default-src 'self'
header: x-webkit-csp: default-src 'self'
header: referrer-policy: same-origin
header: location: https://mapfiles.nrel.gov/data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv
header: vary: Accept
header: Age: 0
header: Via: https/1.1 api-umbrella (ApacheTrafficServer [cMsSf ])
header: X-Cache: MISS
header: Access-Control-Allow-Origin: *
header: Access-Control-Allow-Methods: POST, PUT, DELETE, GET, OPTIONS
header: Access-Control-Request-Method: *
header: Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
header: Strict-Transport-Security: max-age=31536000; preload
DEBUG:urllib3.connectionpool:https://developer.nrel.gov:443 "GET /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-76.2002669+43.0402873%29&interval=60&api_key=API_KEY&email=EMAIL&utc=false HTTP/1.1" 302 125


DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): mapfiles.nrel.gov:443
send: b'GET /data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv HTTP/1.1\r\nHost: mapfiles.nrel.gov\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n'
reply: 'HTTP/1.1 301 Moved Permanently\r\n'
header: Date: Tue, 27 Jul 2021 09:49:50 GMT
header: Content-Type: text/html
header: Content-Length: 169
header: Connection: keep-alive
header: Server: nginx/1.19.7
header: Location: https://gds-hsds-nsrdb-files.nrelcloud.org/data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv
header: Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
DEBUG:urllib3.connectionpool:https://mapfiles.nrel.gov:443 "GET /data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv HTTP/1.1" 301 169


DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): gds-hsds-nsrdb-files.nrelcloud.org:443
send: b'GET /data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv HTTP/1.1\r\nHost: gds-hsds-nsrdb-files.nrelcloud.org\r\nUser-Agent: python-requests/2.25.1\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n'
reply: 'HTTP/1.1 200 OK\r\n'
header: Content-Type: text/csv
header: Content-Length: 465682
header: Connection: keep-alive
header: Date: Tue, 27 Jul 2021 09:39:30 GMT
header: Last-Modified: Tue, 27 Jul 2021 09:39:28 GMT
header: x-amz-expiration: expiry-date="Thu, 29 Jul 2021 00:00:00 GMT", rule-id="24 hour retention"
header: ETag: "5a4d2c1ab0c20acdcbd3b8fd6292edf6"
header: x-amz-storage-class: ONEZONE_IA
header: Server: AmazonS3
header: X-Cache: Hit from cloudfront
header: Via: 1.1 a915962309b732c7a93992ff7088c88d.cloudfront.net (CloudFront)
header: X-Amz-Cf-Pop: PMO50-C1
header: X-Amz-Cf-Id: mG0C3-CY_2IiAa39jF-aD7bxBS5O3srjxklkfvA4HBt8YvtjO77Bmw==
header: Age: 622
DEBUG:urllib3.connectionpool:https://gds-hsds-nsrdb-files.nrelcloud.org:443 "GET /data/solar/1171700_43.05_-76.22_tmy-2020_86eaf692167c07fd77a315d87871e274.csv HTTP/1.1" 200 465682
Downloading file.
Success! File downloaded to /tmp/pysam/nsrdb_43.0402873_-76.2002669_psm3-tmy_60_tmy.csv.

```

</details>

**Comment by dguittet** [comment]
@cpaulgilman Paul, do you know whether this series of steps is required?

**Comment by cpaulgilman** [comment]
The script uses the retry_session function to try the API call up to 10 times until it gets a valid response. In this case, if it didn't retry, API Call 1 would fail and the script would stop, so it seems like the script is doing what we want.

If this is a question about why the API returns 429 twice (API Call 1 and 2), I don't know the answer to that. 

API Call 4 and 5 are necessary because the data was moved, but the API that lists files available for a given lat,lon still points to the old location. That's also not a problem with the script. The extra call to the redirected URL will be necessary until the API is updated.

**Comment by capehart** [comment]
Do all 5 of those calls count against our API rate limit?

**Comment by dguittet** [comment]
This appears to be an effect of your having exceeded the rate limit. When the first query request returns an error, the retry results in the 5 calls you're seeing. When I tried this fresh, I had only 2 calls: one to get the list of files and the second to download. You can see in the below stdout that there are only two `**requests** `

```
test_ResourceTools.py::test_resourcefilefetcher 
Starting data download for solar using 1 thread workers.
Getting list of available NSRDB files for 39.7383155, -105.1800775.

**request** https://developer.nrel.gov/api/nsrdb/v2/solar/nsrdb-data-query.json?api_key=HIiOQJN8C4dcWoJ5cxJ6Cl9LdEOQeo7c2bL67WqA&wkt=POINT(-105.1800775%2039.7383155)

List of available data saved to /Users/dguittet/SAM-Dev/pysam/tests/tmp/nsrdb_data_query_response_39.7383155_-105.1800775.json.
https://developer.nrel.gov/api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-105.1800775+39.7383155%29&interval=60&api_key=HIiOQJN8C4dcWoJ5cxJ6Cl9LdEOQeo7c2bL67WqA&email=dguittet@nrel.gov&utc=false

**request** https://developer.nrel.gov/api/nsrdb/v2/solar/nsrdb-data-query.json?api_key=HIiOQJN8C4dcWoJ5cxJ6Cl9LdEOQeo7c2bL67WqA&wkt=POINT(-105.1800775%2039.7383155)

Downloading file.
Success! File downloaded to /Users/dguittet/SAM-Dev/pysam/tests/tmp/nsrdb_39.7383155_-105.1800775_psm3-tmy_60_tmy.csv.
```

**Comment by ghazi-git** [comment]
@dguittet if you check the API calls log attached in the first comment (it is collapsed by default), you should see the value of the rate limit headers
```
header: X-RateLimit-Limit: 1000
header: X-RateLimit-Remaining: 996
```
in case you want to reproduce the same logs, you can do that by following the steps described [here](https://stackoverflow.com/a/16630836)

**Comment by dguittet** [comment]
I added the logging and to download an NSRDB file this is what I get:

```
Starting data download for solar using 1 thread workers.
Getting list of available NSRDB files for 39.75, -105.22.
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): developer.nrel.gov:443
DEBUG:urllib3.connectionpool:https://developer.nrel.gov:443 "GET /api/nsrdb/v2/solar/nsrdb-data-query.json?api_key=HIiOQJN8C4dcWoJ5cxJ6Cl9LdEOQeo7c2bL67WqA&wkt=POINT(-105.22%2039.75) HTTP/1.1" 200 None
List of available data saved to /Users/dguittet/SAM-Dev/pysam/data/PySAM Downloaded Weather Files/nsrdb_data_query_response_39.75_-105.22.json.
DEBUG:urllib3.connectionpool:https://developer.nrel.gov:443 "GET /api/nsrdb/v2/solar/psm3-tmy-download.csv?names=tmy&wkt=POINT%28-105.22+39.75%29&interval=60&api_key=HIiOQJN8C4dcWoJ5cxJ6Cl9LdEOQeo7c2bL67WqA&email=dguittet@nrel.gov&utc=false HTTP/1.1" 302 125
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): mapfiles.nrel.gov:443
DEBUG:urllib3.connectionpool:https://mapfiles.nrel.gov:443 "GET /data/solar/145808_39.73_-105.22_tmy-2020_151da437360f6bce0c775cb80b64e8e8.csv HTTP/1.1" 301 169
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): gds-hsds-nsrdb-files.nrelcloud.org:443
DEBUG:urllib3.connectionpool:https://gds-hsds-nsrdb-files.nrelcloud.org:443 "GET /data/solar/145808_39.73_-105.22_tmy-2020_151da437360f6bce0c775cb80b64e8e8.csv HTTP/1.1" 200 463861
Downloading file.
Success! File downloaded to /Users/dguittet/SAM-Dev/pysam/data/PySAM Downloaded Weather Files/nsrdb_39.75_-105.22_psm3-tmy_60_tmy.csv.
```

There are 3 HTTPS connections. As @cpaulgilman said, the last two calls are needed because the data has been moved so the 3 I see here seem correct.

**Comment by ghazi-git** [comment]
@dguittet not sure if you're using a temp api key or not but I think it's best to remove it from the logs so it doesn't get abused :) 
I'll be sure to test again and get back to you soon (maybe it's specific to the API Key I'm using).
